File: 윤상우/Week_2/B17298_오큰수.java
package temp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class B17298_오큰수 {

    static int n;
    static String [] arr;
    static Stack<Integer> ans;
    static Stack<Integer> temp;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        n = Integer.parseInt(st.nextToken());
        ans = new Stack<>();
        temp = new Stack<>();

        arr = br.readLine().split(" ");

        for(int i=n-1; i>=0; i--){
            int num = Integer.parseInt(arr[i]);
            while(!temp.isEmpty() && num >= temp.peek()){
                temp.pop();
            }
            if(temp.isEmpty()){
                ans.add(-1);
            }else{
                ans.add(temp.peek());
            }
            temp.add(num);
        }


        while(!ans.isEmpty()){
            sb.append(ans.pop() + " ");
        }

        System.out.println(sb);
    }
}

File: 서지수/Week_2/B17298_오큰수.py
import sys
input=sys.stdin.readline

N=int(input())
A=(list(map(int, input().split())))

result = [-1] * N
stack = []


stack.append(0)
for i in range(1, N):
    while stack:
        if A[stack[-1]] >= A[i]:
            break
        result[stack.pop()] = A[i]
    stack.append(i)

for i in range(len(result)):
    print(result[i], end=" ")

print()